src/OpenTelemetry/Metrics/MetricReader.cs

```diff
@@ -27,35 +28,74 @@ namespace OpenTelemetry.Metrics
     /// </summary>
     public abstract partial class MetricReader : IDisposable
     {
-        private const AggregationTemporality AggregationTemporalityUnspecified = (AggregationTemporality)0;
+        private const MetricReaderTemporalityPreference MetricReaderTemporalityPreferenceUnspecified = (MetricReaderTemporalityPreference)0;
+
+        private static Func<Type, AggregationTemporality> cumulativeTemporatlityPreferenceFunc =
+            (instrumentType) => AggregationTemporality.Cumulative;
+
+        private static Func<Type, AggregationTemporality> monotonicDeltaTemporatlityPreferenceFunc = (instrumentType) =>
+        {
+            return instrumentType.GetGenericTypeDefinition() switch
+            {
+                var type when type == typeof(Counter<>) => AggregationTemporality.Delta,
+                var type when type == typeof(ObservableCounter<>) => AggregationTemporality.Delta,
+                var type when type == typeof(Histogram<>) => AggregationTemporality.Delta,
+
+                // Temporatlity is not defined for gauges, so this does not really affect anything.
+                var type when type == typeof(ObservableGauge<>) => AggregationTemporality.Delta,
+
+                // With .NET 7 the OpenTelemetry .NET SDK will support UpDownCounters.
+                // These will be aggregated using Cumulative temporatlity.
+                // See:
+                //    https://docs.microsoft.com/dotnet/api/system.diagnostics.metrics.updowncounter-1
+                //    https://docs.microsoft.com/dotnet/api/system.diagnostics.metrics.observableupdowncounter-1
+                // var type when type == typeof(UpDownCounter<>) => AggregationTemporality.Cumulative,
+                // var type when type == typeof(ObservableUpDownCounter<>) => AggregationTemporality.Cumulative,
+
+                // TODO: Consider logging here because we should not fall through to this case.
+                _ => AggregationTemporality.Delta,
+            };
+        };
+
         private readonly object newTaskLock = new();
         private readonly object onCollectLock = new();
         private readonly TaskCompletionSource<bool> shutdownTcs = new();
-        private AggregationTemporality temporality = AggregationTemporalityUnspecified;
+        private MetricReaderTemporalityPreference temporalityPreference = MetricReaderTemporalityPreferenceUnspecified;
+        private Func<Type, AggregationTemporality> temporatlityFunc = cumulativeTemporatlityPreferenceFunc;
         private int shutdownCount;
         private TaskCompletionSource<bool> collectionTcs;
         private BaseProvider parentProvider;
 
-        public AggregationTemporality Temporality
+        public MetricReaderTemporalityPreference TemporalityPreference
```

Comment:
To make sure I understand this correctly:
This PR is not exposing the feature to pick Temporality based on Instrument Type to end user. (it can be added in future, as spec requires it).

In a nutshell, This PR replaces Delta/Cumulative with DeltaForMonotonic/Cumulative, along with some renamings.


@alanwest could you confirm if the above understanding is correct?

Reply:
> This PR is not exposing the feature to pick Temporality based on Instrument Type.

Correct. We can introduce fine-grained controls later if we really find there's a need - i.e., expose the ability for the end user to individually choose temporality for each instrument kind. Though, I kinda suspect that will never be a need.

Though in the future we may expand `MetricReaderTemporalityPreference` with the `Stateless` value that has been proposed.

> This PR replaces Delta/Cumulative with DeltaForMonotonic/Cumulative, along with some renamings.

The top-level setting on `MetricReader` for AggregationTemporality (cumulative/delta) is replaced with the new `MetricReaderTemporalityPreference` setting. Though the `AggregationTemporality` enumeration does not go away. That is, there are still only two aggregation temporalities.

`MetricReaderTemporalityPreference` simply provides a name for the common mappings of Instrument kind -> temporality.

Reply:
For our purposes, the setting `MetricReaderTemporalityPreference.MonotonicDelta` does not change anything for the existing instrument kinds we support. It mainly stages us up nicely to support things when UpDownCounters are introduced.

Reply:
Okay.. So `AggregationTemporality` remains, and is exposed via `Metric` class, which is available for exporters/exporter authors (and not really meant for end users) to check the temporality of the Metric stream itself.

The newly added `MetricReaderTemporalityPreference` is for end users to configure things.

This is a breaking change for all Exporter authors (There are 3 exporters I am aware of outside of this repo ), so need a strong changelog note.